controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/LoadBalancerMaintainer.java

```diff
@@ -0,0 +1,120 @@
+        this.nameService = nameService;
+    }
+
+    @Override
+    protected void maintain() {
+        // update application object with load balancer information
+        controller().applications().asList().forEach(this::updateApplicationLoadBalancers);
+
+        // Create or update cnames
+        List<Application> applications = controller().applications().asList();
+        applications.forEach(this::registerLoadBalancerEndpoint);
+
+        // Delete removed application rotations
+        // TODO
+    }
+
+    private void updateApplicationLoadBalancers(Application application) {
+        // Get a list of all load balancers for this applications (for all zones and clusters)
+        Map<ZoneId, List<LoadBalancer>> applicationZoneLoadBalancers = new HashMap<>();
+        for (ZoneId zoneId : application.deployments().keySet()) {
+            try {
+                applicationZoneLoadBalancers.put(zoneId, controller().applications().configServer().getLoadBalancers(new DeploymentId(application.id(), zoneId)));
+            } catch (Exception e) {
+                log.log(LogLevel.INFO,
+                        String.format("Got exception fetching load balancers for application: %s, in zone: %s",
+                                      application.id().toShortString(), zoneId.value()),
+                        e);
+            }
+        }
+
+        // store the load balancers on the deployments
+        controller().applications().lockIfPresent(application.id(), lockedApplication -> storeApplicationWithLoadBalancers(lockedApplication, applicationZoneLoadBalancers));
+    }
+
+    private void storeApplicationWithLoadBalancers(LockedApplication lockedApplication, Map<ZoneId, List<LoadBalancer>> loadBalancers) {
+        for (Map.Entry<ZoneId, List<LoadBalancer>> entry : loadBalancers.entrySet()) {
+            Map<ClusterSpec.Id, HostName> loadbalancerClusterMap = entry.getValue().stream()
+                    .collect(Collectors.toMap(lb -> ClusterSpec.Id.from(lb.id()), lb -> HostName.from(lb.hostname())));
+            lockedApplication = lockedApplication.withDeploymentLoadBalancers(entry.getKey(), loadbalancerClusterMap);
+
+        }
+        controller().applications().store(lockedApplication);
+    }
+
+    private void registerLoadBalancerEndpoint(Application application) {
+        for (Map.Entry<ZoneId, Deployment> deploymentEntry : application.deployments().entrySet()) {
+            ZoneId zone = deploymentEntry.getKey();
+            Deployment deployment = deploymentEntry.getValue();
+            for (Map.Entry<ClusterSpec.Id, HostName> loadBalancers : deployment.loadBalancers().entrySet()) {
+                try {
+                    RecordName recordName = RecordName.from(getEndpointName(loadBalancers.getKey(), application.id(), zone));
+                    RecordData recordData = RecordData.fqdn(loadBalancers.getValue().value());
+                    nameService.createCname(recordName, recordData);
```

Comment:
As discussed offline: This should use a combination of `NameService#findRecord` and not assume that implementation of `NameService#createCname` is idempotent.